### ISSUE TITLE
auth() fails to use public key authentication since commit e640833

### ISSUE THREAD

**Issue by toreanderson** [rkitover/net-ssh2#28]
Before commit e640833 (authored by @salva in order to solve issue #22), the code below worked just fine. Note that the user in question can log in non-interactively using a public key (which is unlocked and stored in the agent) using the `ssh` binary just fine.

````perl
#! /usr/bin/perl

use strict;
use warnings;
use Net::SSH2;

my $s = Net::SSH2->new();

$s->connect($ARGV[0]) or die "connect() failed";
print "Connected!\n";

$s->auth(username => $ENV{'LOGNAME'}, password => undef);
die "Error occurred during auth(): " . $s->error() if $s->error();
die "Not authenticated after auth()" if !$s->auth_ok();
print "Authenticated!\n";
````

However after commit e640833 the script fails:

````
$ perl ssh2.pl some-server
Connected!
Error occurred during auth(): -13 at ssh2.pl line 13.
````

This breaks lots of scripts using `Net::SSH2` for me. I noticed it because Fedora has recently upgraded the shipped version. The last working version was 0.53.

**Comment by salva** [comment]
There are two different issues here:

1. `error` is only guaranteed to return a valid value after some other method indicates failure (usually returning `undef`). It works in the same fashion as `$!`.

  For instance, the correct way to write your code is as follows:

  ```perl
  $s->auth(username => $ENV{'LOGNAME'}, password => undef)
    or die "Error occurred during auth(): " . $s->error();
  ```

2. The changes in e640833 exposed a bug on libssh2. I have released a new version ([0.57_03](https://metacpan.org/release/SALVA/Net-SSH2-0.57_03)) that has a work around for that issue.

**Comment by toreanderson** [comment]
Thank you! I can confirm that the fix in 54ae30a67b59cc6d411b7026897e58300a4b030d works well.

Also thanks for pointing out my wrong usage of `error()`.
